src/libraries/NormalizedAmount.sol

```diff
@@ -1,58 +1,137 @@
+/// @dev NormalizedAmount is a utility library to handle token amounts with different decimals
 // SPDX-License-Identifier: Apache 2
 pragma solidity >=0.8.8 <0.9.0;
 
-type NormalizedAmount is uint64;
+struct NormalizedAmount {
+    uint64 amount;
+    uint8 decimals;
+}
 
-using {gt as >, lt as <, sub as -, add as +, div, min, unwrap} for NormalizedAmount global;
+library NormalizedAmountLib {
+    uint8 constant NORMALIZED_DECIMALS = 8;
 
-function gt(NormalizedAmount a, NormalizedAmount b) pure returns (bool) {
-    return NormalizedAmount.unwrap(a) > NormalizedAmount.unwrap(b);
-}
+    error AmountTooLarge(uint256 amount);
+    error NumberOfDecimalsNotEqual(uint8 decimals, uint8 decimalsOther);
+    error AmountUnderflows(uint64 amountA, uint64 amountB);
 
-function lt(NormalizedAmount a, NormalizedAmount b) pure returns (bool) {
-    return NormalizedAmount.unwrap(a) < NormalizedAmount.unwrap(b);
-}
+    function wrap(uint64 amt, uint8 dec) internal pure returns (NormalizedAmount memory) {
+        return NormalizedAmount(amt, dec);
+    }
 
-function sub(NormalizedAmount a, NormalizedAmount b) pure returns (NormalizedAmount) {
-    return NormalizedAmount.wrap(NormalizedAmount.unwrap(a) - NormalizedAmount.unwrap(b));
-}
+    function unwrap(NormalizedAmount memory a) internal pure returns (uint64, uint8) {
```

Comment:
we can get rid of wrap/unwrap and just use the constructor directly

Reply:
i'd prefer to keep the unwrap method given that solidity doesn't have support for destructuring types via pattern matching (just nice to have a single method that cleanly destructures).